### PR TITLE
Fix UMRs indicated by valgrind (x86_64)

### PR DESCRIPTION
**Pull request by peadar**
See issue #453 - The valgrind tool can be used to detect errors such as uninitialized memory reads, etc, at runtime. This is similar to what ASAN does, but does not require compile-time intervention (at the expense of a far greater runtime performance impact)

The `write_validate` function, in conjunction with one of the `mem_validate` functions, is used to ensure that a specific page is readable before actually accessing a specific address from that page (in order to avoid generating a segmentation fault). It does so by writing the first byte of the page to a pipe. If the page is mapped readable, the write succeeds, while if the page is not mapped, or not writeable, then the write will fail with EFAULT or similar.

It seems that there has previously already been issues with ASAN here - rather than using the `write` function in libc,  `syscall` is used in an attempt to circumvent the memory checks from ASAN.

This patch changes the `write_validate` function to take the original pointer, rather than the page-aligned version, and writes from that specific address. Given that we are about to read that value anyway, (probably to retrieve the contents of the stack in order to fulfill some location expression in the dwarf data), the data at that location should actually be initialized, while the data at the page-aligned offset may not be.  This removes UMR errors reported by valgrind, and possibly other tools that understand the semantics of individual numbered system calls invoked via syscall (We can probably actually remove the syscall() variant of the code, but I've left as-is to reduce the amount of churn for now)

While here, I've also fixed an issue where only the first page containing the data accessed by access_mem is validated as part of `write_validate` - `validate_mem`/`write_validate` now take a size argument, and we instead walk through the content of the value to be verified, validating acces every page-size forward of the first address, or, if that exceeds the extent of the object, the last byte in the object. This function is only used with sizeof (unw_word_t) as a size, so the pointers can only possibly straddle a single page boundary, but it's probably clearer to write it for the general case.  (See issue #232)

This addresses the issue for x86_64. Once done, I can do similar for other architectures using the `write_validate()` mechanism

### CODE REVIEW

File: src/x86_64/Ginit.c [hunk 1]
```diff
@@ -142,18 +142,18 @@ write_validate (void *addr)
...
   return ret;
 }
 
-static int (*mem_validate_func) (void *addr, size_t len);
-static int msync_validate (void *addr, size_t len)
+static int (*mem_validate_func) (unw_word_t addr, size_t len);
+static int msync_validate (unw_word_t addr, size_t len)
 {
-  if (msync (addr, len, MS_ASYNC) != 0)
+  if (msync ( (void *)addr, len, MS_ASYNC) != 0)
```
Comment (djwatson) [inline src/x86_64/Ginit.c:156]: It looks like msync needs to be page aligned also?

man msync "EINVAL addr is not a multiple of PAGESIZE"

**Comment by djwatson** [comment]
One comment on msync, but otherwise looks reasonable.  THanks

File: src/x86_64/Ginit.c (hunk 1 above)
Comment (peadar) [reply src/x86_64/Ginit.c]: Ah, I'd even checked the manpage, but missed the reference in the ERRORS section

I have fixed the problem, and added an env var to force the validation function. This allowed me to test it manually at least. If you'd rather I removed env var toggle from the production code, let me know.

**Comment by peadar** [comment]
Thanks for the fast feedback. Updated the patch. If everything else is ok, I can prepare same for other platforms where relevant

File: src/x86_64/Ginit.c [hunk 2]
```diff
@@ -283,21 +310,30 @@ cache_valid_mem(unw_word_t addr)
...
+  if (len == 0)
     return 0;
 
-  if (mem_validate_func ((void *) addr, len) == -1)
+  if (uwn_page_start (addr) == 0)
     return -1;
 
-  cache_valid_mem(addr);
+  unw_word_t lastbyte = addr + (len - 1); // highest addressed byte of data to access
+  for (;;)
```
Comment (kasperk81) [inline src/x86_64/Ginit.c:322]:
```suggestion
  while (1)
```

**Comment by peadar** [comment]
Is there anything else required from me here? I think the CI tests that failed are failing for other commits too, if I read them correctly.
Let me know if I can help.

**Comment by bregma** [comment]
This all seems reasonable.  Bringing it to other targets would be nice.

**Comment by peadar** [comment]
Thanks! I can spend some time covering other affected targets.

**Pull request merged**